### PR TITLE
chore(ci): Upgrade workflows to non-deprecated runtimes

### PR DESCRIPTION
**Pull request by dev-slatto**
# Description of the issue
Hi! 👋🏻

Several of the Github Actions in the repo is using deprecated runtimes for its workflows. 

# Description of changes
_How does this change address the problem?_
I've upgraded the deprecated actions to versions that uses non-deprecated runtimes.

Note that `setup-go@v4` will try to enable caching unless the `cache` input is explicitly set to `false`. I'll be more than happy to make it false if wanted as I see that you already do a lot of cache in the different pipelines.

Open to feedback if there are any other changes wanted done regarding Github Actions 😄

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
_Describe what tests you have done._

# Requirements
_Before commit the code, please do the following steps._
1. Run `make fmt` and `make fmt-sh`
2. Run `make linter`

### CODE REVIEW

**Comment by khanhntd** [comment]
Hi @dev-slatto ,
Thanks for the contribution and your PR contains all the PR from dependandot (e.g [stale](https://github.com/aws/amazon-cloudwatch-agent/pull/707)) so thanks for that, this would help us a lot though.

**Comment by khanhntd** [comment]
Hi @dev-slatto ,
The PR look good to me and there are no concern over it (but no PM is needed for the changes too so good news) when looking the breaking changes of all the Github Action. (including the [slate](https://github.com/actions/stale/releases) action). I agree with you regarding adding `cache` in [setup@v4](https://github.com/actions/setup-go#v4). Even though I saw its mentioned regarding use the key with the [hash of go sum](https://github.com/actions/setup-go#caching-dependency-files-and-build-outputs), I have not seen whether it can applied to cache with different OS. Therefore, if you upgrading it, then we can use the `PR workflow` to validate it.

```
No worries! For the sake of it and to clean up the PR overview I'll add here what this one closes 😄
```
It will automatically close the PR with dependandot when its checking the Github workflow Github action version but  its good to link it too

**Comment by dev-slatto** [comment]
Cool stuff @khanhntd! Do you need anything else from me to this PR or does it look good? 😄 
I see that one test is failing, but it looks like it's related to some bad creds in a EC2 rather than the changes in this PR.

**Comment by khanhntd** [comment]
Yes @dev-slatto , can you help us trying `setup-go@v4` with cache turning on and deleting the `actions/cache@v3` to see if the cache from ` setup-go@v4` will be good with various OS? Agree with the mac-os are credentials thing so you can ignore it, we have aware of the issue and make an attempt to fix it.

**Comment by dev-slatto** [comment]
Hey again @khanhntd 👋🏻 
Latest commit should let the `setup-go@v4` workflow handle dependency cache in the workflows.

**Comment by khanhntd** [comment]
Re-run the workflow to confirm cache is working. In the mean time, you can also fix the merge conflicts.

**Comment by khanhntd** [comment]
For all cached hmm
```
Warning: Restore cache failed: Some specified paths were not resolved, unable to cache dependencies.
```

**Comment by dev-slatto** [comment]
@khanhntd After what I've read it looks like the `cache-dependency-path` actually needs to reference the go.sum file generated for the cache to work. Do we know all of these paths or would we like to then keep the old cache that supports referencing paths instead of the `sum` files?

**Comment by dev-slatto** [comment]
@khanhntd then the conflict should be resolved and the dedicated cache step should be implemented again 😄

File: .github/workflows/clean-aws-resources.yml [hunk 1]
```diff
@@ -15,29 +15,47 @@ jobs:
...
-        uses: aws-actions/configure-aws-credentials@v1
+        uses: aws-actions/configure-aws-credentials@v2
         with:
           role-to-assume: ${{ secrets.TERRAFORM_AWS_ASSUME_ROLE }}
           aws-region: us-west-2
 
       - name: Clean old ami
         run: go run ./tool/clean/clean_ami/clean_ami.go --tags=clean
+
+  clean-old-file-systems:
```
Comment (khanhntd) [inline .github/workflows/clean-aws-resources.yml:30]: You can delete this as a result of this PR https://github.com/aws/amazon-cloudwatch-agent/pull/720

**Review by khanhntd — COMMENTED** [review]
Only 1 small comment before approving it !!!!! Thanks for spending your effort on this !!!

**Comment by dev-slatto** [comment]
All done @khanhntd! 😄 
Thanks for taking the time and giving the feedback!

**Review by khanhntd — APPROVED** [review]
LGTM. Thanks for the contribution !!! Will get another person to review your PR !!

**Review by williazz — APPROVED** [review]

**Pull request merged**